thunder/torch/__init__.py

```diff
@@ -5331,6 +5331,215 @@ def wait(slf) -> None:
+            dtype=_torch_to_thunder_dtype_map[inp.dtype],
+            requires_grad=inp.requires_grad,
+        )
+    elif isinstance(inp, torch.Size):
+        return tuple(inp)
+    elif isinstance(inp, torch.device):
+        return to_device(inp)
+    elif isinstance(inp, torch.dtype):
+        return _torch_to_thunder_dtype_map[inp]
+    elif isinstance(inp, (str, bool, int, float, complex)):
+        return inp
+    else:
+        raise NotImplementedError(f"Unsupported type: {type(inp)}")
+
+
+def augmented_forward_adaptor(sym_op: Callable):
+    def wrapper(*args, **kwargs):
+        from thunder.core.transforms import VJPDual
+
+        out = sym_op(*args, **kwargs)
+        primal = out if isinstance(out, tuple) else (out,)
+        residuals = ((args, kwargs),)
+        return VJPDual(primal, residuals)
+
+    return wrapper
+
+
+def backward_adaptor(torch_func: Callable):
+    def wrapper(saved_for_backward, *grad_output):
+        inp_args, inp_kwargs = saved_for_backward
+        from thunder.core.pytree import tree_flatten, tree_unflatten
+        from builtins import sum
+
+        flat_args, spec = tree_flatten(inp_args)
+        if sum(tree_map(_is_differentiable, grad_output)) == 0:
+            return tree_unflatten(len(flat_args) * [None], spec)
+        if sum(map(_is_differentiable, tree_flatten(inp_kwargs)[0])) != 0:
+            raise NotImplementedError(
+                f"Exception encountered when doing automatic registration for {torch_func} because there is keyword argument that requires gradient, please use manual registration."
+            )
+        from torch._subclasses.fake_tensor import FakeTensorMode
+
+        with FakeTensorMode() as mode:
+            fake_inp_args, fake_inp_kwargs = tree_map(lambda x: _get_fake_arg(x, mode), (inp_args, inp_kwargs))
+            wrapped_torch_func, diff_args = _make_differentiable_wrapper(torch_func, *fake_inp_args, **fake_inp_kwargs)
+            try:
+                _, fake_vjp = torch.autograd.functional.vjp(
+                    wrapped_torch_func, diff_args, v=tree_map(lambda x: _get_fake_arg(x, mode), grad_output)
+                )
+            except Exception as e:
+                msg = f"Exception encountered when doing automatic registration for {torch_func}, please use manual registration: {repr(e)}"
+                raise NotImplementedError(msg) from e
+
+        out_vjp = tree_map(_fake_type_to_thunder, fake_vjp)
+        iter_out_vjp = iter(out_vjp if isinstance(out_vjp, Sequence) else (out_vjp,))
+        fake_vjp = tuple(next(iter_out_vjp) if _is_differentiable(arg) else None for arg in flat_args)
+        return tree_unflatten(fake_vjp, spec)
+
+    return wrapper
+
+
+def meta_adaptor(torch_func: Callable):
+    def wrapper(*args, **kwargs):
+        from thunder.core.pytree import tree_flatten, tree_unflatten
+        from torch._subclasses.fake_tensor import FakeTensorMode
+
+        if kwargs.get("inplace", False) == True:
```

Comment:
pytorch functions could have `out` kwarg, wouldn't it be better to check and error out for this kwarg?
e.g. https://pytorch.org/docs/stable/generated/torch.add.html

Reply:
Great point, thanks~ I'll add some checks